dpnp/backend/extensions/statistics/histogram.cpp

```diff
@@ -0,0 +1,289 @@
+        return {min, max};
+    }
+
+    template <int _Dims, typename dT>
+    size_t get_bin(const sycl::nd_item<_Dims> &,
+                   const dT *val,
+                   const boundsT &) const
+    {
+        uint32_t edges_count = data.size();
+        uint32_t bins_count = edges_count - 1;
+        const auto *bins = data.get_ptr();
+
+        uint32_t bin =
+            std::upper_bound(bins, bins + edges_count, val[0], Less<dT>{}) -
+            bins - 1;
+        bin = std::min(bin, bins_count - 1);
+
+        return bin;
+    }
+
+    template <typename dT>
+    bool in_bounds(const dT *val, const boundsT &bounds) const
+    {
+        Less<dT> _less;
+        return !_less(val[0], std::get<0>(bounds)) &&
+               !_less(std::get<1>(bounds), val[0]) && !IsNan<dT>::isnan(val[0]);
+    }
+
+private:
+    DataStorage data;
+};
+
+template <typename T>
+using CachedEdges = HistogramEdges<T, CachedData<const T, 1>>;
+
+template <typename T>
+using UncachedEdges = HistogramEdges<T, UncachedData<const T, 1>>;
+
+template <typename T, typename BinsT, typename HistType = size_t>
+struct histogram_kernel
+{
+    static sycl::event impl(sycl::queue &exec_q,
+                            const void *vin,
+                            const void *vbins_edges,
+                            const void *vweights,
+                            void *vout,
+                            const size_t bins_count,
+                            const size_t size,
+                            const std::vector<sycl::event> &depends)
+    {
+        const T *in = static_cast<const T *>(vin);
+        const BinsT *bins_edges = static_cast<const BinsT *>(vbins_edges);
+        const HistType *weights = static_cast<const HistType *>(vweights);
+        HistType *out = static_cast<HistType *>(vout);
+
+        auto device = exec_q.get_device();
+        const auto local_size = get_max_local_size(device);
+
+        constexpr uint32_t WorkPI = 128; // empirically found number
+
+        const auto nd_range = make_ndrange(size, local_size, WorkPI);
+
+        return exec_q.submit([&](sycl::handler &cgh) {
+            cgh.depends_on(depends);
+            constexpr uint32_t dims = 1;
+
+            auto dispatch_edges = [&](uint32_t local_mem, auto &weights,
+                                      auto &hist) {
+                if (device.is_gpu() && (local_mem >= bins_count + 1)) {
+                    auto edges = CachedEdges(bins_edges, bins_count + 1, cgh);
+                    submit_histogram(in, size, dims, WorkPI, hist, edges,
+                                     weights, nd_range, cgh);
+                }
+                else {
+                    auto edges = UncachedEdges(bins_edges, bins_count + 1, cgh);
+                    submit_histogram(in, size, dims, WorkPI, hist, edges,
+                                     weights, nd_range, cgh);
+                }
+            };
+
+            auto dispatch_bins = [&](auto &weights) {
```

Comment:
Looks like this could be `const auto &weights`.

Reply:
done